### PR TITLE
feat: historacle proto files

### PR DESCRIPTION
**Pull request by adamewozniak**
## Description

historacle protobuf files

This is a small PR so that next I can do:

1. Types
2. Keeper functions
3. Migrations :^) 

relates to #1521

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/umee/historacle/v1/tx.proto [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import "cosmos/msg/v1/msg.proto";
+
+option go_package = "github.com/umee-network/umee/v3/x/historacle/types";
+option (gogoproto.goproto_getters_all) = false;
+
+// Msg defines the historacle Msg service.
+service Msg {
+  // StampPeriod defines the interface for which governance
+  // can set the stamp period.
+  rpc UpdateStampPeriod(MsgUpdateStampPeriod)
```
Comment (adamewozniak) [inline proto/umee/historacle/v1/tx.proto:15]:
```suggestion
  // UpdateStampPeriod defines the interface which governance
  // can use to set the stamp period.
  rpc UpdateStampPeriod(MsgUpdateStampPeriod)
```

File: proto/umee/historacle/v1/query.proto [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  option (gogoproto.equal)           = false;
+  option (gogoproto.goproto_getters) = false;
+  // denom defines the denomination to query for.
+  string denom = 1;
+}
+
+// QueryMediansResponse is response type for the
+// Query/Medians RPC method.
+message QueryMediansResponse {
+  // exchange_rates defines a list of the medians for all stamped denoms.
```
Comment (toteki) [inline proto/umee/historacle/v1/query.proto:33]:
```suggestion
  // medians defines a list of the medians for all stamped denoms.
```

File: proto/umee/historacle/v1/historacle.proto [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+// HistoricPrice defines the object used to stamp prices.
+message HistoricPrice {
+  repeated umee.oracle.v1.ExchangeRateTuple exchange_rates = 1 [
+        (gogoproto.castrepeated) = "ExchangeRateTuples",
+        (gogoproto.nullable)     = false,
+        (gogoproto.moretags) = "yaml:\"exchange_rates\""
+      ];
+  uint64 block_num = 2 [
+    (gogoproto.moretags) = "yaml:\"block_num\""
```
Comment (toteki) [inline proto/umee/historacle/v1/historacle.proto:27]: Could rename to `height`

**Review by toteki — APPROVED** [review]

**Review by robert-zaremba — DISMISSED** [review]
Blocking. IMO it should be part of x/oracle module.
Adding new module is an overhead in the code and maintainance.  

Let's move it, unless there is a valid reason to not keep it in x/oracle.

File: proto/umee/historacle/v1/tx.proto [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+message MsgUpdateStampPeriod {
+  option (gogoproto.equal)           = false;
+  option (gogoproto.goproto_getters) = false;
+  option (cosmos.msg.v1.signer) = "authority";
+
+  string         authority    = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string         title        = 2 [(gogoproto.moretags) = "yaml:\"title\""];
+  string         description  = 3 [(gogoproto.moretags) = "yaml:\"description\""];
+  uint64         stamp_period = 4 [(gogoproto.moretags) = "yaml:\"stamp_period\""];
+  uint64         prune_period = 5 [(gogoproto.moretags) = "yaml:\"prune_period\""];
```
Comment (robert-zaremba) [inline proto/umee/historacle/v1/tx.proto:30]: `yaml` tags are not supported any more in the compiler, so let's remove them.

**Comment by robert-zaremba** [comment]
let's generate the proto files in the same PR, otherwise it can potentially conflict or make harder with anyone else who can work on proto.

**Comment by adamewozniak** [comment]
> Blocking. IMO it should be part of x/oracle module. Adding new module is an overhead in the code and maintainance.
> 
> Let's move it, unless there is a valid reason to not keep it in x/oracle.

Imo it's a bit cleaner to keep it as a separate module (one for the tallying, verifying, and acceptance of prices, and another for the record keeping of those prices), but I don't really mind either way

**Comment by adamewozniak** [comment]
gotta fix the lint

**Comment by adamewozniak** [comment]
Actually now I'm leaning towards keeping it in a separate module. We plan to have a release before we come out with price history and it'll get messy; with a separate module we could just not include it in app.go. 

Thoughts? Cc @toteki @robert-zaremba @RafilxTenfen

File: x/oracle/keeper/grpc_query.go [hunk 5]
```diff
@@ -249,3 +249,10 @@ func (q querier) AggregateVotes(
 		AggregateVotes: votes,
 	}, nil
 }
+
+// Medians currently performs a no-op.
+func (q querier) Medians(
+	_ context.Context, _ *types.QueryMedians,
+) (*types.QueryMediansResponse, error) {
+	return &types.QueryMediansResponse{}, nil
```
Comment (toteki) [inline x/oracle/keeper/grpc_query.go:257]: Creating an `ErrNotImplemented` would be good here rather than returning empty response but `nil` error

**Comment by toteki** [comment]
After our discussion about the other historical data we want it to be able to support in the future, I'm on board with making it a separate module so the oracle module stays lean

**Comment by brandoncomer** [comment]
Just want to give some context. Historacle should be its own module since we plan to add quite a bit of functionality to it going forward (e.g. 30 day moving averages, etc.). This will be necessary when we spin off to Ojo.

**Comment by adamewozniak** [comment]
Yeah I think I'm gonna move this back over lol

image

**Comment by adamewozniak** [comment]
doin some cleanup, feel free to add comments

File: proto/umee/historacle/v1/historacle.proto [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+option go_package = "github.com/umee-network/umee/v3/x/historacle/types";
+
+option (gogoproto.goproto_getters_all) = false;
+
+// Params defines the parameters for the historacle module.
+message Params {
+  option (gogoproto.equal)            = true;
+  option (gogoproto.goproto_stringer) = false;
+  uint64 stamp_period                 = 1;
```
Comment (zarazan) [inline proto/umee/historacle/v1/historacle.proto:15]: Maybe add a comment to define what stamp_period and prune_period are?

File: proto/umee/historacle/v1/tx.proto [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    // StampPeriod defines the interface for which governance
+    // can use to set the stamp period.
+    rpc UpdateStampPeriod(MsgUpdateStampPeriod)
+        returns (MsgUpdateStampPeriodResponse);
+  }
+
+
+// MsgUpdateStampPeriod represents a message for governance to update the
+// stamp period.
+message MsgUpdateStampPeriod {
```
Comment (zarazan) [inline proto/umee/historacle/v1/tx.proto:21]: Is is possible we would want to add a second stamp period (ex: 7 day median and 30 day median). nit: It slightly bothers me  that this is called `MsgUpdateStampPeriod` and updates the stamp_period and prune_period.

Comment (adamewozniak) [reply proto/umee/historacle/v1/tx.proto:21]: I think for now we can keep it as a single stamp period, but yeah this should be renamed

Comment (adamewozniak) [reply proto/umee/historacle/v1/tx.proto]: 15b70efb8230123a7e58873b98da247fec2872ea

File: proto/umee/historacle/v1/historacle.proto (hunk 6 above)
Comment (adamewozniak) [reply proto/umee/historacle/v1/historacle.proto:16]: 15b70efb8230123a7e58873b98da247fec2872ea

**Review by zarazan — APPROVED** [review]
nice.

**Comment by robert-zaremba** [comment]
Module can be fat... Moreover, we can add new proto without enabling it in the next release. This is a standard practice, and in a long run we will have a not finished feature while we will need to make a release.

Also for a note: go module != SDK module != unit of functionality.

Let's discuss on Slack, will be easier.

File: proto/umee/historacle/v1/genesis.proto [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+option go_package = "github.com/umee-network/umee/v3/x/historacle/types";
+
+option (gogoproto.goproto_getters_all) = false;
+
+// GenesisState defines the historacle module's genesis state.
+message GenesisState {
+  Params                               params          = 1 [(gogoproto.nullable) = false];
+  repeated cosmos.base.v1beta1.DecCoin medians         = 2 [(gogoproto.nullable) = false];
+  repeated HistoricPrice               historic_prices = 3 [(gogoproto.nullable) = false];
```
Comment (robert-zaremba) [inline proto/umee/historacle/v1/genesis.proto:16]: how about renaming to `prices`

**Comment by adamewozniak** [comment]
> Module can be fat... Moreover, we can add new proto without enabling it in the next release. This is a standard practice, and in a long run we will have a not finished feature while we will need to make a release.
> 
> Also for a note: go module != SDK module != unit of functionality.
> 
> Let's discuss on Slack, will be easier.

Honestly I'm ok with this either way, going to wait to get some more input, but I don't mind merging this back into x/oracle

File: proto/umee/historacle/v1/genesis.proto (hunk 8 above)
Comment (adamewozniak) [reply proto/umee/historacle/v1/genesis.proto:16]: If we're keeping this in the oracle module I think `historic_prices` is more clear

**Comment by adamewozniak** [comment]
alright, switched this back
cc: @robert-zaremba @zarazan

**Review by zarazan — APPROVED** [review]

**Comment by robert-zaremba** [comment]
> Moved back into the oracle module, need to move on this

Thank you. Looks much leaner now. BTW - you could ping me on Slack.

**Pull request merged**